### ISSUE TITLE
OperationCanceledException handling in ReadUntilCancelledAsync() and ReadAllAsEnumerablesAsync()

### ISSUE THREAD

**Issue by mrak1990** [Open-NET-Libraries/Open.ChannelExtensions#40]
Now ReadUntilCancelledAsync() and ReadAllAsEnumerablesAsync() catches all OperationCanceledException exceptions, even if they were thrown by user code in receiver.

```csharp
using var httpClient = new HttpClient();
httpClient.Timeout = TimeSpan.FromSeconds(5);
var url = "https://reqres.in/api/users?delay=10"; // responds with a delay of 10 seconds

async ValueTask LoadAndPrint()
{
    var json = await httpClient.GetStringAsync(url, ct); // throws TaskCanceledException due to timeout
    Console.WriteLine(json[..10]);
}

// must throw TaskCanceledException
_ = await Enumerable.Repeat(0, 10)
    .ToChannel(cancellationToken: ct)
    .ReadUntilCancelledAsync(ct, (_, _) => LoadAndPrint());

// must throw TaskCanceledException
await Enumerable.Repeat(0, 10)
    .ToChannel(cancellationToken: ct)
    .ReadAllAsEnumerablesAsync(_ => LoadAndPrint(), cancellationToken: ct);
```

**Comment by electricessence** [comment]
I don't think this is going to change, because it's intentional to "just bail" if there's a cancellation token that has been cancelled for these specific methods.
The token itself is effectively "owned/borrowed" by the caller, and instead of a try catch, you simply call `token.ThrowIfCancellationRequested()`.

Take a look at the differences between the other various methods.  It's intentional.  There are plenty of cases where it absolutely will throw when the token is cancelled because that's expected for something as simple as `.ReadAllAsync()` or the like.

And again, for `.ReadAllAsEnumerables[Async]`: "cancellelation" is a 'signal' to **_bail out_**.  And then the caller can or should own what they want to do with it.

**Comment by mrak1990** [comment]
> The token itself is effectively "owned/borrowed" by the caller, and instead of a try catch, you simply call `token.ThrowIfCancellationRequested()`.

I'm not sure I got the idea right.
In the example above, the exception `TaskCanceledException` (inherited from `OperationCanceledException`) is thrown inside the `HttpClient.GetStringAsync()` method due to a timeout. At the same time, the ct token is not cancelled.
As a result, `ReadAllAsEnumerablesAsync()` is executed without exception (although not the whole pipeline is worked out) and `ct.ThrowIfCancellationRequested();` will not throw an exception.

**Comment by electricessence** [comment]
I think I see the nuanced edge case here...
You are suggesting that if the delegate itself throws a cancellation exception, but the CancellationToken is not cancelled, then it should not trap the exception.

**Comment by electricessence** [comment]
Fixed: https://www.nuget.org/packages/Open.ChannelExtensions/8.4.0

https://github.com/Open-NET-Libraries/Open.ChannelExtensions/commit/018def45d8ab78adda0faa3af7b5fa1edf475381

**Comment by electricessence** [comment]
I will also update the minor versions for 6 and 7.

**Comment by electricessence** [comment]
Another few changes to make things a bit more correct:
https://www.nuget.org/packages/Open.ChannelExtensions/8.4.1

If you'd be so kind as to review these changes too:
https://github.com/Open-NET-Libraries/Open.ChannelExtensions/commit/b1f4df088aa22f030b65471baf6f2006477f9e66

**Comment by mrak1990** [comment]
LGTM
As I understand, the changes to `ReadUntilCancelledAsync`/`ReadAllAsEnumerablesAsync` (`WaitToReadOrCancelAsync -> WaitToReadAsync`) change the behaviour on explicit cancellation. Now these methods throw `OperationCancelledException` if the token passed above in chain is canceled.
This makes sense, now the behavior is similar to the methods from BCL.

**Comment by mrak1990** [comment]
I found another example of incorrect behavior when throwing `OperationCancelledException`.
```csharp
ValueTask<int> AsyncReceiver()
{
    throw new OperationCanceledException("test exception");
}

// OK, throw exception
_ = await Enumerable.Repeat(0, 10)
    .ToChannel()
    .ReadAllAsync(async _ => await AsyncReceiver());

// NOT OK, doesn't throw exception
_ = await Enumerable.Repeat(0, 10)
    .ToChannel()
    .PipeAsync(1, async _ => await AsyncReceiver())
    .ReadAll(_ => { });
```

As I understand it, the problem is how `ContinueWith` handles `OperationCancelledException` (`t.Exception = null` and `t.Status = Cancelled` [here](https://github.com/Open-NET-Libraries/Open.ChannelExtensions/blob/master/Open.ChannelExtensions/Extensions.Pipe.cs#L101))

**Comment by electricessence** [comment]
> LGTM

The update is already in for the previous examples.
I'll need some time to investigate the one  you just posted.

**Comment by electricessence** [comment]
@mrak1990 
I strongly suggest in the future, if you have a fork and want to merge a PR, start with adding a test to the test project that proves the current behavior is incorrect.  Then I can pull that down and decide what to do and ultimately merge the fix.

**Comment by electricessence** [comment]
Investigating now.

**Comment by electricessence** [comment]
> As I understand it, the problem is how `ContinueWith` handles `OperationCancelledException` (`t.Exception = null` and `t.Status = Cancelled` [here](https://github.com/Open-NET-Libraries/Open.ChannelExtensions/blob/master/Open.ChannelExtensions/Extensions.Pipe.cs#L101))

You are correct.  This one is interesting.

**Comment by electricessence** [comment]
Fixed and released to 8.4.2.  Great catch.  Thank you.

Added special and simple handling that if the task is cancelled, but not by the token, then be sure to tell the channel why by passing it an OperationCancelledException.
https://github.com/Open-NET-Libraries/Open.ChannelExtensions/blob/master/Open.ChannelExtensions/Extensions.Pipe.cs#L11

**Comment by electricessence** [comment]
I'll update the v6 and v7 versions soon.
